lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+  for (unsigned i = 0; i < NumElements; i++) {
+    CloseProximityRelations[i].resize(NumElements);
+    for (unsigned j = 0; j < NumElements; j++) {
+      CloseProximityRelations[i][j] = calculateCloseProximity(i, j);
+      if (i < j && CloseProximityRelations[i][j] > MaxWCP) {
+        MaxWCP = CloseProximityRelations[i][j];
+        MaxI = i;
+        MaxJ = j;
+      }
+    }
+  }
+  if (MaxWCP < 1e-3) {
+    Eligibility = false;
+    return;
+  }
+  DEBUG(dbgs() << "CPG for field reordering recommendation\n");
+  for (unsigned i = 0; i < NumElements; i++) {
+    DEBUG(dbgs() << "F" << i + 1 << "\t");
+    for (unsigned j = 0; j < NumElements; j++)
+      DEBUG(dbgs() << DEBUG_PRINT_COUNT(CloseProximityRelations[i][j]) << "\t");
+    DEBUG(dbgs() << "\n");
+  }
+
+  // FIXME: Need to reconsider if the order of the pair makes differences
+  NewOrder.push_back(MaxI);
+  NewOrder.push_back(MaxJ);
+  FieldsToTransform.erase(MaxI);
+  FieldsToTransform.erase(MaxJ);
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER, dbgs() << "Use F" << MaxI + 1 << " and F"
+                                             << MaxJ + 1
+                                             << " as seeds in reordering\n");
+  assert(MaxWCP == getWCP());
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "Current WCP: " << DEBUG_PRINT_COUNT(MaxWCP)
+                         << "\n");
+}
+
+// Calculate Close Proximity in this reordering algorithm. This algorithm
+// takes in a pair of CP relation of (count, distance) and return a single
+// floating point number that is count/distance. Special cases is that when
+// two fields are the same, or they can't fit into the same cache block,
+// return 0. When distance is smaller than 1 byte, return the count as if
+// the distance is 1 byte.
+double
+FieldReorderAnalyzer::calculateCloseProximity(FieldNumType Field1,
+                                              FieldNumType Field2) const {
+  if (Field1 == Field2)
+    return 0;
+  auto *Pair = CPBuilder->getCloseProximityPair(Field1, Field2);
+  if (Pair->first < 1e-3)
+    return 0;
+  if (Pair->second < 1)
+    return Pair->first;
+  return Pair->first / Pair->second;
+}
+
+// Calculate weighted CP (WCP) value of all fields within a cache block
+// The function takes an array of fields that can fit on this cache block and
+// returns a sum of CP value between every pair of fields on the block
+double FieldReorderAnalyzer::calculateWCPWithinCacheBlock(
+    std::vector<FieldNumType> *CacheBlock) const {
+  double WCP = 0;
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "Calculate WCP within a cache block: ");
+  for (unsigned i = 0; i < CacheBlock->size(); i++) {
+    DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                    dbgs() << "F" << (*CacheBlock)[i] + 1 << " ");
+    for (unsigned j = i + 1; j < CacheBlock->size(); j++)
+      if ((*CacheBlock)[i] < (*CacheBlock)[j])
+        WCP += CloseProximityRelations[(*CacheBlock)[i]][(*CacheBlock)[j]];
```

Comment:
I'll remove this checking. It's redundant. Whether I use the array or a getter, I don't need this checking anyway.